lib/src/results.dart

```diff
@@ -207,3 +210,33 @@ extension RealmResultsInternal on RealmResults {
     return RealmResults<T>._(handle, realm);
   }
 }
+
+class _RealmResultsIterator<T extends RealmObject> implements Iterator<T> {
+  final RealmResults<T> _results;
+  final int _length;
+  int _index;
+  T? _current;
+
+  _RealmResultsIterator(RealmResults<T> results)
+      : _results = results,
+        _length = results.length,
+        _index = 0;
+
+  @override
+  T get current => _current as T;
+
+  @override
+  bool moveNext() {
+    int length = _results.length;
+    if (_length != length) {
+      throw ConcurrentModificationError(_results);
+    }
+    if (_index >= length) {
+      _current = null;
+      return false;
+    }
+    _current = _results[_index];
+    _index++;
```

Comment:
I don't like that `_index` and `_current` are pointing to different elements in the collection - it's not super intuitive and took me a while to convince myself that the code is correct. Even if we do keep the eager evaluation of `_current`, I would strongly suggest that we modify the code to ensure `_index` and `_current` are consistent.

Reply:
We could start `_index` on -1, and increment before setting `_current`

Reply:
This would allow us to implement `T get current => _results[_index]`, but I still tend to prefer an explicit `_current` that is set in `moveNext`.. just for the odd chance that collection is changed after `moveNext`, but before calling `current`

Reply:
It should not change assuming both calls happen on the same thread without yielding the thread between them, but as I said, I don't have strong opinions on the subject since the assumption is that `moveNext` and `current` will always be called as a pair. The reason we don't do eager evaluation in .NET is because it makes the code more concise and we're lazy 😝

Reply:
It is my opinion, but let's leave this PR for removeMany like it is. The implementation of the iterator will be changed for sure later. We have other issues also that we need to take in consideration: allowing nullable objects in results, editing or not objects while iterate, snapshot or not, concurrently changing objects order for sorted results and changing results collection for filtered results. We probably will implement common iterator for RealmList and RealmResults. I can prepare a list of test cases. It is important to have a decision how apps developers would expect it to behave. Let's discuss it with Lubo first.

Reply:
I tend to agree with @nirinchev. And we need to shapshot the collection.

Reply:
I changed index to start from -1, so that now current and index are at the same point as @nirinchev required. I kept having _current member and evaluation in advance as @nielsenko suggested. Snapshot will be done with issue #166 - discussed with @blagoev. If you agree I would resolve this conversation.